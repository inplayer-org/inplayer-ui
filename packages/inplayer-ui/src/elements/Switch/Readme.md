Toggle is a input component for flipswitch. You can use it directly like this:

```jsx static
import { Label, Switch } from '@inplayer-org/inplayer-ui';

class Page extends React.Component {
   state = {
    checked: false,
  };

  onToggleChange = checked => {
    this.setState({
      checked,
    });
  };

  return <Switch checked={value} id="toggle1" label="swipe" onChange={this.onToggleChange} />
}
```

### Demo

```jsx
<Switch
  disabled={false}
  checked={false}
  id="unchecked"
  label="Swipe me!"
  onChange={console.log}
/>
<br />
<Switch
  disabled={false}
  checked={true}
  id="checkedByDefault"
  label="Checked by default!"
  onChange={console.log}
/>
<br />
<Switch
  disabled={true}
  checked={false}
  id="disabledUnchecked"
  label="Disabled!"
  onChange={console.log}
/>
<br />
<Switch
  disabled={true}
  checked={true}
  id="disabledChecked"
  label="Disabled and checked!"
  onChange={console.log}
/>
```
