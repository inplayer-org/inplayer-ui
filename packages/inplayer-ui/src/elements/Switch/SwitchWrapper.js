import styled from 'styled-components';
import uiColors from 'utils/uiColors';
import colors from 'config/colors';
import { transparentize } from 'polished';

const SwitchWrapper = styled.span`
  display: inline-flex;
  align-items: center;
  justify-content: center;

  .react-toggle-thumb {
    transition: all 0.5s cubic-bezier(0.23, 1, 0.32, 1) 0ms;
    position: absolute;
    top: -1px;
    left: 0;
    width: 22px;
    height: 22px;
    border-radius: 50%;
    border: 1px solid ${colors.gray};
    background: ${colors.white};
    box-sizing: border-box;
  }

  .react-toggle-track {
    width: 34px;
    height: 18px;
    padding: 0;
    border-radius: 30px;
    background-color: ${colors.lightGray};
    transition: all 0.2s ease;
    border: 1px solid ${colors.gray};
  }

  .react-toggle {
    touch-action: pan-x;
    display: inline-block;
    position: relative;
    cursor: pointer;
    background-color: transparent;
    border: 0;
    padding: 0;
    user-select: none;
  }

  .react-toggle-screenreader-only {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -0.0625rem;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }

  .react-toggle--checked .react-toggle-track {
    border: 1px solid ${uiColors('primary.main')};
    background: ${uiColors('primary.main')};
  }

  &:hover .react-toggle-track {
    background: ${transparentize(0.94, colors.skyBlue)};
    border: 1px solid ${uiColors('primary.main')};
  }

  .react-toggle.react-toggle--disabled {
    pointer-events: none;

    .react-toggle-track {
      background: ${transparentize(0.94, colors.black)} !important;
      border: 1px solid ${colors.gray} !important;
    }
  }

  .react-toggle--checked .react-toggle-thumb {
    left: 14px;
  }

  &:hover .react-toggle--disabled .react-toggle-track .react-toggle-thumb {
    border: none;
  }

  &:hover .react-toggle:not(.react-toggle--disabled) .react-toggle-thumb {
    border: 1px solid ${uiColors('primary.main')};
  }

  .react-toggle-track-check {
    position: absolute;
    width: 14px;
    height: 10px;
    top: 0;
    bottom: 0;
    margin-top: auto;
    margin-bottom: auto;
    line-height: 0;
    left: 8px;
    opacity: 0;
    transition: opacity 0.25s ease;
  }

  .react-toggle--checked .react-toggle-track-check {
    opacity: 1;
    transition: opacity 0.25s ease;
  }

  .react-toggle-track-x {
    position: absolute;
    width: 10px;
    height: 10px;
    top: 0;
    bottom: 0;
    margin-top: auto;
    margin-bottom: auto;
    line-height: 0;
    right: 10px;
    opacity: 1;
    transition: opacity 0.25s ease;
  }

  .react-toggle--checked .react-toggle-track-x {
    opacity: 0;
  }
`;

/** @component */
export default SwitchWrapper;
