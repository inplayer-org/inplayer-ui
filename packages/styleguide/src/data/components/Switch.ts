import { ComponentData } from "../types";

const usage = `
  import { Label, Switch } from '@inplayer-org/inplayer-ui';

  class Page extends React.Component {
    state = {
      checked: false,
    };

    onToggleChange = checked => {
      this.setState({
        checked,
      });
    };

    return <Switch checked={value} id="toggle1" label="swipe" onChange={this.onToggleChange} />
  }
`;

const demoCode = `
<div>
  <Switch
  disabled={false}
  checked={false}
  id="unchecked"
  label="Swipe me!"
  onChange={console.log}
  />
  <br />
  <Switch
  disabled={false}
  checked={true}
  id="checkedByDefault"
  label="Checked by default!"
  onChange={console.log}
  />
  <br />
  <Switch
  disabled={true}
  checked={false}
  id="disabledUnchecked"
  label="Disabled!"
  onChange={console.log}
  />
  <br />
  <Switch
  disabled={true}
  checked={true}
  id="disabledChecked"
  label="Disabled and checked!"
  onChange={console.log}
/>
</div>
`;

const Switch: ComponentData = {
  description: `Toggle is a input component for flipswitch. You can use it directly like this:`,
  path: `src/elements/Switch/index.js`,
  usage,
  examples: [
    {
      title: 'Demo',
      code: demoCode,
    }
  ],
  propsAndMethods: [
    { name: 'checked', type: 'boolean', default: 'Required' },
    { name: 'id', type: 'string', default: 'Required' },
    { name: 'label', type: 'string', default: 'Required' },
    { name: 'onChange', type: 'function', default: 'Required' },
    {name: 'className', type: 'string', description: 'A className can be passed down for further styling or extending with CSS-in-JS'},
    { name: 'disabled', type: 'boolean', default: 'false' },
    { name: 'style', type:'Object', default: '{}' },
  ]
}

export default Switch;