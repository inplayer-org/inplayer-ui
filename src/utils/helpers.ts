import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
