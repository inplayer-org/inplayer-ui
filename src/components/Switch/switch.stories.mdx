import { useState } from 'react';
import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Switch from './'

<Meta title="Switch" component={<Switch/>} />

# Switch


```jsx
import { Switch } from '@inplayer-org/inplayer-ui';

const Page = () => (
    <Switch />
)
```

<Props of={Switch}/>

### Standard
<Preview withSource="open">
  <Story name="standard">
    {() => {
        const [checked, setChecked] = useState(false);
        return (
            <Switch
                checked={checked}
                id="toggle1"
                label="Swipe me!"
                onChange={() => setChecked(!checked)}
            />
        )
    }}
  </Story>
</Preview>

### Checked
<Preview withSource="open">
  <Story name="checked">
    {() => {
        const [checked, setChecked] = useState(true);
        return (
            <Switch
                checked={checked}
                id="toggle1"
                label="Checked by default!"
                onChange={() => setChecked(!checked)}
            />
        )
    }}
  </Story>
</Preview>

### Disabled and not checked
<Preview withSource="open">
  <Story name="disable - not checked">
    {() => {
        const [checked, setChecked] = useState(false);
        return (
            <Switch
                checked={checked}
                id="toggle1"
                label="Disabled and not checked!"
                onChange={() => setChecked(!checked)}
                disabled
            />
        )
    }}
  </Story>
</Preview>

### Disabled and checked
<Preview withSource="open">
  <Story name="disabled - checked">
    {() => {
        const [checked, setChecked] = useState(true);
        return (
            <Switch
                checked={checked}
                id="toggle1"
                label="Disabled and checked!"
                onChange={() => setChecked(!checked)}
                disabled
            />
        )
    }}
  </Story>
</Preview>

