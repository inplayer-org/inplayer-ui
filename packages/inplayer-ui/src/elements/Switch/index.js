// @flow
import React from 'react';
import styled from 'styled-components';
import Toggle from 'react-toggle';
import SwitchWrapper from './SwitchWrapper';
import Label from '../Label';

type Props = {
  checked: boolean,
  disabled?: boolean,
  id: string,
  label: string,
  onChange: (checked: boolean) => any,
  /** A className can be passed down for further styling or extending with CSS-in-JS */
  className?: string,
  style?: Object,
};

const MarginLeftLabel = styled(Label)`
  margin-left: 0.5rem;
`;

const Switch = ({ id, checked, label, disabled, onChange, className, style }: Props) => {
  const onToggleChange = (e: SyntheticInputEvent<*>) => {
    if (e.target) {
      onChange(e.target.checked);
    }
  };
  return (
    <SwitchWrapper className={className} style={style}>
      <Toggle
        id={id}
        disabled={disabled}
        checked={checked}
        onChange={onToggleChange}
        icons={false}
      />
      <MarginLeftLabel disabled={disabled} htmlFor={id}>
        {label}
      </MarginLeftLabel>
    </SwitchWrapper>
  );
};

Switch.defaultProps = {
  className: '',
  style: {},
  disabled: false,
};

/** @component */
export default Switch;
