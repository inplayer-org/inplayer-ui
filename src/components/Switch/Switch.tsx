import React, { ChangeEvent } from 'react';
import styled from 'styled-components';
import Toggle from 'react-toggle';
import SwitchWrapper from './SwitchWrapper';
import Label from '../Label';

type SwitchProps = {
  /**
   * Whether the switch is checked or not
   */
  checked: boolean;
  /**
   * Disable switch
   */
  disabled?: boolean;
  /**
   * Id
   */
  id: string;
  /**
   * Switch label
   */
  label: string;
  /**
   * Function executed when the switch value changes
   */
  onChange: (checked: boolean) => void;
};

const MarginLeftLabel = styled(Label)`
  margin-right: 0.5rem;
`;

const Switch = ({ id, checked, label, disabled = false, onChange }: SwitchProps) => {
  const onToggleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (e.target) {
      onChange(e.target.checked);
    }
  };

  return (
    <SwitchWrapper>
      <MarginLeftLabel disabled={disabled} htmlFor={id}>
        {label}
      </MarginLeftLabel>
      <Toggle
        id={id}
        disabled={disabled}
        checked={checked}
        onChange={onToggleChange}
        icons={false}
      />
    </SwitchWrapper>
  );
};

export default Switch;
