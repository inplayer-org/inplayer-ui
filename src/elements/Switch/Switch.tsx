
import React, { ChangeEvent, CSSProperties } from 'react';
import styled  from 'styled-components';
import Toggle from 'react-toggle';
import SwitchWrapper from './SwitchWrapper';
import Label from '../Label';

interface Props {
  checked: boolean;
  disabled?: boolean;
  id: string;
  label: string;
  onChange: (checked: boolean) => void;
  className?: string;
  style?: CSSProperties;
}

const MarginLeftLabel = styled(Label)<{ disabled?: boolean }>`
  margin-left: 0.5rem;
`;

const Switch = ({ id, checked, label, disabled = false, onChange, className = '', style = {} }: Props) => {
  const onToggleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (e.target) {
      onChange(e.target.checked);
    }
  };
  return (
    <SwitchWrapper className={className} style={style}>
      <Toggle
        id={id}
        disabled={disabled}
        checked={checked}
        onChange={onToggleChange}
        icons={false}
      />
      <MarginLeftLabel disabled={disabled} htmlFor={id}>
        {label}
      </MarginLeftLabel>
    </SwitchWrapper>
  );
};




export default Switch;
