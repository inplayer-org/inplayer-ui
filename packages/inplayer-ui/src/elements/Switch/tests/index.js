import { generateSnapshots } from 'packages/inplayer-ui/src/elements/Switch/tests/helpers';

import Switch from '../index';

generateSnapshots(Switch);
