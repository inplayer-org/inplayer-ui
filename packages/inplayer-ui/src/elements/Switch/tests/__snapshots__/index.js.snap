// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders correctly 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 .react-toggle-thumb {
  -webkit-transition: all 0.5s cubic-bezier(0.23,1,0.32,1) 0ms;
  transition: all 0.5s cubic-bezier(0.23,1,0.32,1) 0ms;
  position: absolute;
  top: -1px;
  left: 0;
  width: 22px;
  height: 22px;
  border-radius: 50%;
  border: 1px solid #d7dde5;
  background: #fff;
  box-sizing: border-box;
}

.c0 .react-toggle-track {
  width: 34px;
  height: 18px;
  padding: 0;
  border-radius: 30px;
  background-color: #F6F8FB;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  border: 1px solid #d7dde5;
}

.c0 .react-toggle {
  touch-action: pan-x;
  display: inline-block;
  position: relative;
  cursor: pointer;
  background-color: transparent;
  border: 0;
  padding: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c0 .react-toggle-screenreader-only {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -0.0625rem;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c0 .react-toggle--checked .react-toggle-track {
  border: 1px solid #07AAE6;
  background: #07AAE6;
}

.c0:hover .react-toggle-track {
  background: rgba(7,170,230,0.06);
  border: 1px solid #07AAE6;
}

.c0 .react-toggle.react-toggle--disabled {
  pointer-events: none;
}

.c0 .react-toggle.react-toggle--disabled .react-toggle-track {
  background: rgba(0,0,0,0.06) !important;
  border: 1px solid #d7dde5 !important;
}

.c0 .react-toggle--checked .react-toggle-thumb {
  left: 14px;
}

.c0:hover .react-toggle--disabled .react-toggle-track .react-toggle-thumb {
  border: none;
}

.c0:hover .react-toggle:not(.react-toggle--disabled) .react-toggle-thumb {
  border: 1px solid #07AAE6;
}

.c0 .react-toggle-track-check {
  position: absolute;
  width: 14px;
  height: 10px;
  top: 0;
  bottom: 0;
  margin-top: auto;
  margin-bottom: auto;
  line-height: 0;
  left: 8px;
  opacity: 0;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c0 .react-toggle--checked .react-toggle-track-check {
  opacity: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c0 .react-toggle-track-x {
  position: absolute;
  width: 10px;
  height: 10px;
  top: 0;
  bottom: 0;
  margin-top: auto;
  margin-bottom: auto;
  line-height: 0;
  right: 10px;
  opacity: 1;
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c0 .react-toggle--checked .react-toggle-track-x {
  opacity: 0;
}

.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 300;
  cursor: pointer;
  vertical-align: middle;
  color: #3e4348;
  -webkit-transition: all ease 300ms;
  transition: all ease 300ms;
  line-height: 20px;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-left: 0.5rem;
}

.c1:not(:disabled):hover {
  color: #7d7e7f;
}

<span
  className="c0"
  style={Object {}}
>
  <div
    className="react-toggle"
    onClick={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="react-toggle-track"
    >
      <div
        className="react-toggle-track-check"
      />
      <div
        className="react-toggle-track-x"
      />
    </div>
    <div
      className="react-toggle-thumb"
    />
    <input
      className="react-toggle-screenreader-only"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
  </div>
  <label
    className="c1"
    disabled={false}
  />
</span>
`;
